querybook/webapp/components/AIAssistant/QueryGenerationModal.tsx

```diff
@@ -0,0 +1,233 @@
+import React, { useEffect, useState } from 'react';
+
+import { QueryEngineSelector } from 'components/QueryRunButton/QueryRunButton';
+import { QueryComparison } from 'components/TranspileQueryModal/QueryComparison';
+import { IQueryEngine } from 'const/queryEngine';
+import { StreamStatus, useStream } from 'hooks/useStream';
+import { TableToken } from 'lib/sql-helper/sql-lexer';
+import { matchKeyPress } from 'lib/utils/keyboard';
+import { analyzeCode } from 'lib/web-worker';
+import { Button } from 'ui/Button/Button';
+import { DebouncedInput } from 'ui/DebouncedInput/DebouncedInput';
+import { Icon } from 'ui/Icon/Icon';
+import { Message } from 'ui/Message/Message';
+import { Modal } from 'ui/Modal/Modal';
+import { StyledText } from 'ui/StyledText/StyledText';
+
+import { AIMode, AIModeSelector } from './AIModeSelector';
+import { TableSelector } from './TableSelector';
+
+import './QueryGenerationModal.scss';
+
+interface IProps {
+    dataCellId: number;
+    query: string;
+    engineId: number;
+    queryEngines: IQueryEngine[];
+    queryEngineById: Record<number, IQueryEngine>;
+    onUpdateQuery?: (query: string) => void;
+    onUpdateEngineId: (engineId: number) => void;
+    onHide: () => void;
+}
+
+const useTablesInQuery = (query, language) => {
+    const [tables, setTables] = useState<string[]>([]);
+
+    useEffect(() => {
+        analyzeCode(query, 'autocomplete', language).then((codeAnalysis) => {
+            const tableReferences: TableToken[] = [].concat.apply(
+                [],
+                Object.values(codeAnalysis?.lineage.references ?? {})
+            );
+            setTables(
+                tableReferences.map(({ schema, name }) => `${schema}.${name}`)
+            );
+        });
+    }, [query, language]);
+
+    return tables;
+};
+
+export const QueryGenerationModal = ({
+    dataCellId,
+    query = '',
+    engineId,
+    queryEngines,
+    queryEngineById,
+    onUpdateQuery,
+    onUpdateEngineId,
+    onHide,
+}: IProps) => {
+    const inputRef = React.useRef<HTMLInputElement>(null);
+    const tablesInQuery = useTablesInQuery(
+        query,
+        queryEngineById[engineId]?.language
+    );
+    const [question, setQuestion] = useState<string>('');
+    const [tables, setTables] = useState(tablesInQuery);
+    const [aiMode, setAIMode] = useState(
+        !!query ? AIMode.EDIT : AIMode.GENERATE
+    );
+
+    useEffect(() => {
+        setTables([...new Set([...tablesInQuery, ...tables])]);
```

Comment:
use https://lodash.com/docs/3.10.1#uniq with tables.concat(tablesInQuery), so the order isn't shuffled